tsl/src/continuous_aggs/materialize.c

```diff
@@ -492,25 +519,74 @@ get_materialization_end_point_for_table(int32 raw_hypertable_id, int32 materiali
 	return end_time;
 }
 
-static int64
-hypertable_get_min(SchemaAndName hypertable, Name time_column, Oid time_type, bool *found)
+/* get min and max value for timestamp column for hypertable */
+static bool
+hypertable_get_min_and_max_time_value(SchemaAndName hypertable, Name time_column,
+									  int64 search_start, Oid time_type, int64 *min_out,
+									  int64 *max_out)
 {
-	Datum min_time_datum;
-	bool val_is_null;
+	Datum last_time_value;
+	Datum first_time_value;
+	bool fval_is_null, lval_is_null;
+	bool search_start_is_infinite = false;
+	bool found_new_tuples = false;
 	StringInfo command = makeStringInfo();
 	int res;
-	int64 min_time_internal = PG_INT64_MIN;
-	*found = false;
+
+	Datum search_start_val =
+		internal_to_time_value_or_infinite(search_start, time_type, &search_start_is_infinite);
+
+	if (search_start_is_infinite && search_start > 0)
+	{
+		/* the previous completed time was +infinity, there can be no new ranges */
+		return false;
+	}
 
 	res = SPI_connect();
 	if (res != SPI_OK_CONNECT)
 		elog(ERROR, "could not connect to SPI while search for new tuples");
 
-	appendStringInfo(command,
-					 "SELECT min(%s) FROM %s.%s",
-					 quote_identifier(NameStr(*time_column)),
-					 quote_identifier(NameStr(*hypertable.schema)),
-					 quote_identifier(NameStr(*hypertable.name)));
+	/* We always SELECT both max and min in the following queries. There are two cases
+	 * 1. there is no index on time: then we want to perform only one seqscan.
+	 * 2. there is a btree index on time: then postgres will transform the query
+	 *    into two index-only scans, which should add very little extra work
+	 *    compared to the materialization.
+	 * Ordered append append also fires, so we never scan beyond the first and last chunks
+	 */
+	if (search_start_is_infinite)
+	{
+		/* previous completed time is -infinity, or does not exist, so we must scan from the
+		 * beginning */
+		appendStringInfo(command,
+						 "SELECT max(%s), min(%s) FROM %s.%s",
+						 quote_identifier(NameStr(*time_column)),
+						 quote_identifier(NameStr(*time_column)),
+						 quote_identifier(NameStr(*hypertable.schema)),
+						 quote_identifier(NameStr(*hypertable.name)));
+	}
+	else
+	{
+		Oid out_fn;
+		bool type_is_varlena;
+		char *search_start_str;
+
+		getTypeOutputInfo(time_type, &out_fn, &type_is_varlena);
+
+		search_start_str = OidOutputFunctionCall(out_fn, search_start_val);
+
+		*min_out = search_start;
+		/* normal case, add a WHERE to take advantage of chunk constraints */
+		/*We handled the +infinity case above*/
+		Assert(!search_start_is_infinite);
+		appendStringInfo(command,
+						 "SELECT max(%s), min(%s) FROM %s.%s WHERE %s >= %s",
```

Comment:
The statement is identical with the other case until `WHERE` clause. Can it be moved out the if and only `WHERE` condition appended here?